### PR TITLE
Add examples: 1-getting-started.ipynb and 2-multi-step.ipynb

### PR DESCRIPTION
**Pull request by mromaniukcdl**
Created a directory for examples, moved the existing example there and added a new one (1-getting-started.ipynb)

### CODE REVIEW

File: examples/1-getting-started.ipynb [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn.datasets import load_digits\n",
+    "\n",
+    "digits = load_digits()\n",
+    "X_digits, y_digits = digits.data, digits.target\n",
+    "\n",
+    "num_train = int(0.8 * len(y_digits))\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl I usually take advantage of sklearn.model_selection module with either simple `train_test_split` (as here) or `StratifiedKFold` . Is there any reason not to go with 
`X_train, X_valid, y_train, y_valid = train_test_split(X_digits, y_digits, test_size=0.2)` ?

File: examples/1-getting-started.ipynb [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    "                }\n",
+    "            }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating steps\n",
+    "Let's create a simple step - first we define an **adapter** to tell it how to interpret its input data dictionary (this allows you to do lots of clever things but we'll stick to basics for now)"
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl as I understand you introduce `adapter` because running this without the `adapter` doesn't work. If so we should fix/implement simple option right away to be able make shift without this new concept in the very first tutorial.
Can you confirm that running without the `adapter` doesn't work?

File: examples/1-getting-started.ipynb [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    "                     'X': [('input', 'X')],\n",
+    "                     'y': [('input', 'y')]\n",
+    "                 }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The second ingredient of a step is a transformer, which is where the real action happens. You just have to define a class following a **simple API** and then it's up to you to be as creative as you want!\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl I really like your style. It reads really nicely!

File: examples/1-getting-started.ipynb [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    "## Training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "classifier_step.clean_cache()\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl I wouldn't introduce `.clean_cache()` just yet. Let's leave this for the future (notebook explaining `cache_output`, `save_output` and stuff)

File: examples/1-getting-started.ipynb [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    "preds_train = classifier_step.fit_transform(data_train);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "acc_train = np.sum(preds_train['y_pred'] == data_train['input']['y']) / data_train['input']['y'].size"
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl I think going with the `sklearn.metrics.accuracy_score` is even more readable.

File: examples/1-getting-started.ipynb [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    "preds_test = classifier_step.transform(data_test)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "acc_test = np.sum(preds_test['y_pred'] == data_test['input']['y']) / data_test['input']['y'].size\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl same here. I would just go with the sklearn

File: examples/1-getting-started.ipynb (hunk 1 above)
Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: I was trying to do something along the lines of this example: http://scikit-learn.org/stable/auto_examples/exercises/plot_digits_classification_exercise.html#sphx-glr-auto-examples-exercises-plot-digits-classification-exercise-py

StratifiedKFold would be better in principle although a few more lines of code. But we could switch to it.

Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: +1

File: examples/1-getting-started.ipynb (hunk 4 above)
Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: +1 in principle but I think we would like to turn off caching altogether in this introductory tutorial

File: examples/1-getting-started.ipynb (hunk 5 above)
Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: +1

File: examples/1-getting-started.ipynb (hunk 6 above)
Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: +1

File: examples/1-getting-started.ipynb [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "fix, axs = plt.subplots(4, 8, figsize=(10, 6))\n",
+    "for ii, ax in enumerate(axs.ravel()):\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl why ii instaed of i ?

File: examples/1-getting-started.ipynb [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    "At this point it may seem like a lot of work for not much benefit but once we start moving towards more complex pipelines, the reasoning behind all the components will become more clear. Have a look at the next notebook for a more advanced, multi-step pipeline!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# TODO: Get rid of this\n",
```
Comment (jakubczakon) [inline examples/1-getting-started.ipynb]: @mromaniukcdl Todo should not be in a PR

File: examples/2-multi-step.ipynb [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+    "from steps.base import Step, BaseTransformer"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# TODO: Get rid of this\n",
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl again Todo should not be in a PR

File: examples/2-multi-step.ipynb [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+    "        self.mean = np.mean(X, axis=0)\n",
+    "        self.std = np.std(X, axis=0)\n",
+    "        return self\n",
+    "\n",
+    "    def transform(self, X, **kwargs):\n",
+    "        X_tfm  = (X - self.mean) / self.std\n",
+    "        return {'X': X_tfm}\n",
+    "    \n",
+    "    def save(self, filepath):\n",
+    "        joblib.dump([self.mean, self.std], filepath)\n",
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl I would go with tuple instead of list since those are not elements of the same "type".

File: examples/2-multi-step.ipynb [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+    "    def load(self, filepath):\n",
+    "        self.estimator = joblib.load(filepath)\n",
+    "        return self"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Finally, we'll use logistic regression as our classifier:"
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl at the beginning RandomForest was promised

File: examples/2-multi-step.ipynb [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    "    def __init__(self):\n",
+    "        self.estimator = LogisticRegression()\n",
+    "        \n",
+    "    def fit(self, X, y):\n",
+    "        self.estimator.fit(X, y)\n",
+    "        return self\n",
+    "\n",
+    "    def transform(self, X, **kwargs):\n",
+    "        y_pred  = self.estimator.predict(X)\n",
+    "        return {'y_pred': y_pred}  # TODO: exaplain this\n",
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl I would try and not submit Todos

File: examples/2-multi-step.ipynb [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+    "Training a pipeline is a one-liner:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "preds_train = rf_step.fit_transform(data_train);  # TODO: It's weird that we're fitting the final step. I get the idea but it still seems unintuitive that the last step represents the whole pipeline..."
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl fair point. We should discuss it at the next planning.

File: examples/2-multi-step.ipynb [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+    "That seems pretty good. Have a look at the next notebook for even more complex pipelines with parallel branches."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# TODO: Get rid of this\n",
```
Comment (jakubczakon) [inline examples/2-multi-step.ipynb]: @mromaniukcdl Todo to drop

File: examples/2-multi-step.ipynb (hunk 13 above)
Comment (kamil-kaczmarek) [reply examples/2-multi-step.ipynb]: @mromaniukcdl @jakubczakon I think it is just a matter of explanation that you run `pipeline(data)`

File: examples/1-getting-started.ipynb [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    "This is just about the simplest pipeline you can imagine. Now let's train it!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training\n",
+    "\n",
+    "Training a pipeline is a one-liner:"
```
Comment (kamil-kaczmarek) [inline examples/1-getting-started.ipynb]: @jakubczakon @mromaniukcdl I think that some explanation that we do `pipeline(data)` and why it works is needed here.

**Comment by kamil-kaczmarek** [comment]
(@jakubczakon) @mromaniukcdl thanks for this notebook! I just have some general level comments:
* we may want to add more info at the beginning, such like what `Steps` are, why we build them, that why they are useful.
* Should we link to code in the markdown cells of the notebook?
* I really like the idea to have final comments about what you have learned/showed in the notebook! Maybe we can further improve this with discussion

File: examples/1-getting-started.ipynb (hunk 7 above)
Comment (mromaniukcdl) [reply examples/1-getting-started.ipynb]: Matlab convention. Will change it.

File: examples/1-getting-started.ipynb [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+    "* **save** and **load** its parameters; this is handy when you're trying to avoid re-computing things over and over.\n",
+    "\n",
+    "See how flexible this is? You can just as easily wrap your Keras or Pytorch models."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's turn our adapter into a step:"
```
Comment (grzes314) [inline examples/1-getting-started.ipynb]: > Now let's turn our adapter into a step:

Did you mean transformer?

File: examples/2-multi-step.ipynb [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+    "            }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating pipeline components\n",
+    "\n",
+    "This time we want to make a smarter attempt at digit classification. We'll normalize our data, run PCA to compute some features of a different flavour and then combine them with our original features in a final Random Forest step."
```
Comment (grzes314) [inline examples/2-multi-step.ipynb]: But you use breast cancer data set, not digits.

Comment (grzes314) [reply examples/2-multi-step.ipynb]: Also as far as I see you use logistic regression, not random forests.

File: examples/2-multi-step.ipynb [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+    "Our classifier step will have to combine two data flows: the features processed by PCA, and the labels fed directly from input. Therefore, we will have to use the `adapter` argument to specify how to map those inputs to transformer arguments."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "rf_step = Step(name='LogReg',\n",
```
Comment (grzes314) [inline examples/2-multi-step.ipynb]: `rf_step` might be a confusing name for logistic regression step.

File: examples/1-getting-started.ipynb (hunk 15 above)
Comment (jakubczakon) [reply examples/1-getting-started.ipynb]: @kamil-kaczmarek @mromaniukcdl fair. I would go with one line saying steps are fitted recursively so executing fit_transform or transform on data effectively fits the entire pipeline.

File: .gitignore [hunk 19]
```diff
@@ -1,3 +1,4 @@
 .idea/
 .ipynb_checkpoints/
+steps/__pycache__/
```
Comment (grzes314) [inline .gitignore]: I would add `*.pyc` instead of `__pycache__`
